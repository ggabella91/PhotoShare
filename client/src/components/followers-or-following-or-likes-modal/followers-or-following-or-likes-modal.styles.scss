.followers-following-modal {
  display: flex;
  flex-direction: column;
  justify-content: center;
  height: 50vh;
  width: 50vw;
  min-width: 260px;
  max-width: 400px;
  min-height: 400px;
  max-height: 400px;
  margin-top: 25vh;
  margin-left: 50%;
  transform: translate(-50%);

  .modal-content {
    width: 50vw;
    max-width: inherit;
    min-width: inherit;
    max-height: 400px;
    border-radius: 15px;
  }

  .followers-following-modal-header {
    display: flex;
    justify-content: center;
    position: relative;

    .header-text {
      margin-left: 50%;
      transform: translate(-50%);
      font-size: 18px;
      font-weight: 420;
      text-align: center;
    }
  }

  .followers-following-modal-body {
    display: flex;
    flex-direction: column;
    min-height: 338px;
    overflow-y: scroll;
    overflow-x: hidden;
    padding: 0;
  }
}
