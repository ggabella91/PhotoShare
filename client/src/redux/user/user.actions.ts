import {
  UserActions,
  UserActionTypes,
  UserSignUp,
  UserSignIn,
  User,
  FieldsToUpdate,
  OtherUserRequest,
  ChangePassword,
  ResetPassword,
  Error,
} from './user.types';

export const signUpStart = ({
  username,
  name,
  email,
  password,
  passwordConfirm,
}: UserSignUp): UserActionTypes => ({
  type: UserActions.SIGN_UP_START,
  payload: { username, name, email, password, passwordConfirm },
});

export const signUpSuccess = (user: User): UserActionTypes => ({
  type: UserActions.SIGN_UP_SUCCESS,
  payload: user,
});

export const signUpFailure = (error: Error): UserActionTypes => ({
  type: UserActions.SIGN_UP_FAILURE,
  payload: error,
});

export const checkUserSession = (): UserActionTypes => ({
  type: UserActions.CHECK_USER_SESSION,
  payload: null,
});

export const setCurrentUser = (user: User | null): UserActionTypes => ({
  type: UserActions.SET_CURRENT_USER,
  payload: user,
});

export const signInStart = ({
  email,
  password,
}: UserSignIn): UserActionTypes => ({
  type: UserActions.SIGN_IN_START,
  payload: {
    email,
    password,
  },
});

export const signInSuccess = (user: User): UserActionTypes => ({
  type: UserActions.SIGN_IN_SUCCESS,
  payload: user,
});

export const signInFailure = (error: Error): UserActionTypes => ({
  type: UserActions.SIGN_IN_FAILURE,
  payload: error,
});

export const signOutStart = (): UserActionTypes => ({
  type: UserActions.SIGN_OUT_START,
  payload: null,
});

export const signOutSuccess = (): UserActionTypes => ({
  type: UserActions.SIGN_OUT_SUCCESS,
  payload: null,
});

export const signOutFailure = (error: Error): UserActionTypes => ({
  type: UserActions.SIGN_OUT_FAILURE,
  payload: error,
});

export const changeInfoStart = (
  fieldsToUpdate: FieldsToUpdate
): UserActionTypes => ({
  type: UserActions.CHANGE_INFO_START,
  payload: fieldsToUpdate,
});

export const changeInfoSuccess = (user: User): UserActionTypes => ({
  type: UserActions.CHANGE_INFO_SUCCESS,
  payload: user,
});

export const changeInfoFailure = (error: Error): UserActionTypes => ({
  type: UserActions.CHANGE_INFO_FAILURE,
  payload: error,
});

export const changePasswordStart = ({
  passwordCurrent,
  password,
  passwordConfirm,
}: ChangePassword): UserActionTypes => ({
  type: UserActions.CHANGE_PASSWORD_START,
  payload: { passwordCurrent, password, passwordConfirm },
});

export const changePasswordSuccess = (message: string): UserActionTypes => ({
  type: UserActions.CHANGE_PASSWORD_SUCCESS,
  payload: message,
});

export const changePasswordFailure = (error: Error): UserActionTypes => ({
  type: UserActions.CHANGE_PASSWORD_FAILURE,
  payload: error,
});

export const forgotPasswordStart = (email: string): UserActionTypes => ({
  type: UserActions.FORGOT_PASSWORD_START,
  payload: email,
});

export const forgotPasswordSuccess = (message: string): UserActionTypes => ({
  type: UserActions.FORGOT_PASSWORD_SUCCESS,
  payload: message,
});

export const forgotPasswordFailure = (error: Error): UserActionTypes => ({
  type: UserActions.FORGOT_PASSWORD_FAILURE,
  payload: error,
});

export const resetPasswordStart = ({
  password,
  passwordConfirm,
  token,
}: ResetPassword): UserActionTypes => ({
  type: UserActions.RESET_PASSWORD_START,
  payload: { password, passwordConfirm, token },
});

export const resetPasswordSuccess = (message: string): UserActionTypes => ({
  type: UserActions.RESET_PASSWORD_SUCCESS,
  payload: message,
});

export const resetPasswordFailure = (error: Error): UserActionTypes => ({
  type: UserActions.RESET_PASSWORD_FAILURE,
  payload: error,
});

export const deleteAccountStart = (): UserActionTypes => ({
  type: UserActions.DELETE_ACCOUNT_START,
  payload: null,
});

export const deleteAccountSuccess = (message: string): UserActionTypes => ({
  type: UserActions.DELETE_ACCOUNT_SUCCESS,
  payload: message,
});

export const deleteAccountFailure = (error: Error): UserActionTypes => ({
  type: UserActions.DELETE_ACCOUNT_FAILURE,
  payload: error,
});

export const clearInfoStatuses = (): UserActionTypes => ({
  type: UserActions.CLEAR_INFO_STATUSES,
  payload: null,
});

export const clearPasswordStatuses = (): UserActionTypes => ({
  type: UserActions.CLEAR_PASSWORD_STATUSES,
  payload: null,
});

export const getOtherUserStart = (
  otherUserReq: OtherUserRequest
): UserActionTypes => ({
  type: UserActions.GET_OTHER_USER_START,
  payload: otherUserReq,
});

export const getOtherUserSuccess = (user: User): UserActionTypes => ({
  type: UserActions.GET_OTHER_USER_SUCCESS,
  payload: user,
});

export const getFollowersSuccess = (user: User): UserActionTypes => ({
  type: UserActions.GET_FOLLOWERS_INFO_SUCCESS,
  payload: user,
});

export const getFollowingSuccess = (user: User): UserActionTypes => ({
  type: UserActions.GET_FOLLOWING_INFO_SUCCESS,
  payload: user,
});

export const getPostReactorsSuccess = (user: User): UserActionTypes => ({
  type: UserActions.GET_POST_REACTOR_SUCCESS,
  payload: user,
});

export const getOtherUserFailure = (error: Error): UserActionTypes => ({
  type: UserActions.GET_OTHER_USER_FAILURE,
  payload: error,
});

export const getUserSuggestionsStart = (match: string): UserActionTypes => ({
  type: UserActions.GET_USER_SUGGESTIONS_START,
  payload: match,
});

export const getUserSuggestionsSuccess = (users: User[]): UserActionTypes => ({
  type: UserActions.GET_USER_SUGGESTIONS_SUCCESS,
  payload: users,
});

export const getUserSuggestionsFailure = (error: Error): UserActionTypes => ({
  type: UserActions.GET_USER_SUGGESTIONS_FAILURE,
  payload: error,
});

export const clearUserSuggestions = (): UserActionTypes => ({
  type: UserActions.CLEAR_USER_SUGGESTIONS,
  payload: null,
});

export const clearFollowersAndFollowing = (): UserActionTypes => ({
  type: UserActions.CLEAR_FOLLOWERS_AND_FOLLOWING,
  payload: null,
});

export const setIsCurrentUserProfilePage = (
  isCurrentUserProfilePage: boolean
): UserActionTypes => ({
  type: UserActions.SET_IS_CURRENT_USER_PROFILE_PAGE,
  payload: isCurrentUserProfilePage,
});

export const getFeedPostReactorsSuccess = (user: User): UserActionTypes => ({
  type: UserActions.GET_FEED_POST_REACTOR_SUCCESS,
  payload: user,
});

export const getConversationUserSuccess = (user: User): UserActionTypes => ({
  type: UserActions.GET_CONVERSATION_USER_SUCCESS,
  payload: user,
});

export const getNotificationUserSuccess = (user: User): UserActionTypes => ({
  type: UserActions.GET_NOTIFICATION_USER_SUCCESS,
  payload: user,
});

export const clearConversationUsers = (): UserActionTypes => ({
  type: UserActions.CLEAR_CONVERSATION_USERS,
});
