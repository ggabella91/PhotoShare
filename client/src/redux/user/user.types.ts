export enum UserActions {
  CHECK_USER_SESSION = 'CHECK_USER_SESSION',
  SET_CURRENT_USER = 'SET_CURRENT_USER',
  SIGN_UP_START = 'SIGN_UP_START',
  SIGN_UP_SUCCESS = 'SIGN_UP_SUCCESS',
  SIGN_UP_FAILURE = 'SIGN_UP_FAILURE',
  SIGN_IN_START = 'SIGN_IN_START',
  SIGN_IN_SUCCESS = 'SIGN_IN_SUCCESS',
  SIGN_IN_FAILURE = 'SIGN_IN_FAILURE',
  SIGN_OUT_START = 'SIGN_OUT_START',
  SIGN_OUT_SUCCESS = 'SIGN_OUT_SUCCESS',
  SIGN_OUT_FAILURE = 'SIGN_OUT_FAILURE',
  CHANGE_INFO_START = 'CHANGE_INFO_START',
  CHANGE_INFO_SUCCESS = 'CHANGE_INFO_SUCCESS',
  CHANGE_INFO_FAILURE = 'CHANGE_INFO_FAILURE',
  CHANGE_PASSWORD_START = 'CHANGE_PASSWORD_START',
  CHANGE_PASSWORD_SUCCESS = 'CHANGE_PASSWORD_SUCCESS',
  CHANGE_PASSWORD_FAILURE = 'CHANGE_PASSWORD_FAILURE',
  FORGOT_PASSWORD_START = 'FORGOT_PASSWORD_START',
  FORGOT_PASSWORD_SUCCESS = 'FORGOT_PASSWORD_SUCCESS',
  FORGOT_PASSWORD_FAILURE = 'FORGOT_PASSWORD_FAILURE',
  RESET_PASSWORD_START = 'RESET_PASSWORD_START',
  RESET_PASSWORD_SUCCESS = 'RESET_PASSWORD_SUCCESS',
  RESET_PASSWORD_FAILURE = 'RESET_PASSWORD_FAILURE',
  DELETE_ACCOUNT_START = 'DELETE_ACCOUNT_START',
  DELETE_ACCOUNT_SUCCESS = 'DELETE_ACCOUNT_SUCCESS',
  DELETE_ACCOUNT_FAILURE = 'DELETE_ACCOUNT_FAILURE',
  CLEAR_INFO_STATUSES = 'CLEAR_INFO_STATUSES',
  CLEAR_PASSWORD_STATUSES = 'CLEAR_PASSWORD_STATUSES',
  GET_OTHER_USER_START = 'GET_OTHER_USER_START',
  GET_OTHER_USER_SUCCESS = 'GET_OTHER_USER_SUCCESS',
  GET_OTHER_USER_FAILURE = 'GET_OTHER_USER_FAILURE',
  CLEAR_OTHER_USER_INFO = 'CLEAR_OTHER_USER_INFO',
}

export interface UserSignUp {
  name: string;
  email: string;
  password: string;
  passwordConfirm: string;
}

export interface UserSignIn {
  email: string;
  password: string;
}

export interface User {
  id: string;
  name: string;
  email: string;
  username: string;
  photo?: string;
  bio?: string;
}

export interface FieldsToUpdate {
  name?: string;
  email?: string;
  username?: string;
  bio?: string;
  photo?: string;
}

export interface ChangePassword {
  passwordCurrent: string;
  password: string;
  passwordConfirm: string;
}

export interface ResetPassword {
  password: string;
  passwordConfirm: string;
  token: string;
}

export interface Error {
  statusCode: number;
  message: string;
}

export type UserPayload =
  | UserSignUp
  | UserSignIn
  | User
  | ChangePassword
  | ResetPassword
  | Error
  | null;

export interface UserState {
  currentUser: User | null;
  otherUser: User | null;
  otherUserError: Error | null;
  signUpError: Error | null;
  signInOrOutError: Error | null;
  changeInfoConfirm: null | string;
  changeInfoError: null | Error;
  changePasswordConfirm: null | string;
  changePasswordError: null | Error;
  forgotConfirm: null | string;
  forgotError: null | Error;
  resetConfirm: null | string;
  resetError: null | Error;
  deleteAccountError: null | Error;
}

export interface SignUpStart {
  type: typeof UserActions.SIGN_UP_START;
  payload: UserSignUp;
}

export interface SignUpSuccess {
  type: typeof UserActions.SIGN_UP_SUCCESS;
  payload: User;
}

export interface SignUpFailure {
  type: typeof UserActions.SIGN_UP_FAILURE;
  payload: Error;
}

export interface CheckUserSession {
  type: typeof UserActions.CHECK_USER_SESSION;
  payload: null;
}

export interface SetCurrentUser {
  type: typeof UserActions.SET_CURRENT_USER;
  payload: User | null;
}

export interface SignInStart {
  type: typeof UserActions.SIGN_IN_START;
  payload: UserSignIn;
}

export interface SignInSuccess {
  type: typeof UserActions.SIGN_IN_SUCCESS;
  payload: User;
}

export interface SignInFailure {
  type: typeof UserActions.SIGN_IN_FAILURE;
  payload: Error;
}

export interface SignOutStart {
  type: typeof UserActions.SIGN_OUT_START;
  payload: null;
}

export interface SignOutSuccess {
  type: typeof UserActions.SIGN_OUT_SUCCESS;
  payload: null;
}

export interface SignOutFailure {
  type: typeof UserActions.SIGN_OUT_FAILURE;
  payload: Error;
}

export interface ChangeInfoStart {
  type: typeof UserActions.CHANGE_INFO_START;
  payload: FieldsToUpdate;
}

export interface ChangeInfoSuccess {
  type: typeof UserActions.CHANGE_INFO_SUCCESS;
  payload: User;
}

export interface ChangeInfoFailure {
  type: typeof UserActions.CHANGE_INFO_FAILURE;
  payload: Error;
}

export interface ChangePasswordStart {
  type: typeof UserActions.CHANGE_PASSWORD_START;
  payload: ChangePassword;
}

export interface ChangePasswordSuccess {
  type: typeof UserActions.CHANGE_PASSWORD_SUCCESS;
  payload: string;
}

export interface ChangePasswordFailure {
  type: typeof UserActions.CHANGE_PASSWORD_FAILURE;
  payload: Error;
}

export interface ForgotPasswordStart {
  type: typeof UserActions.FORGOT_PASSWORD_START;
  payload: string;
}

export interface ForgotPasswordSuccess {
  type: typeof UserActions.FORGOT_PASSWORD_SUCCESS;
  payload: string;
}

export interface ForgotPasswordFailure {
  type: typeof UserActions.FORGOT_PASSWORD_FAILURE;
  payload: Error;
}

export interface ResetPasswordStart {
  type: typeof UserActions.RESET_PASSWORD_START;
  payload: ResetPassword;
}

export interface ResetPasswordSuccess {
  type: typeof UserActions.RESET_PASSWORD_SUCCESS;
  payload: string;
}

export interface ResetPasswordFailure {
  type: typeof UserActions.RESET_PASSWORD_FAILURE;
  payload: Error;
}

export interface DeleteAccountStart {
  type: typeof UserActions.DELETE_ACCOUNT_START;
  payload: null;
}

export interface DeleteAccountSuccess {
  type: typeof UserActions.DELETE_ACCOUNT_SUCCESS;
  payload: string;
}

export interface DeleteAccountFailure {
  type: typeof UserActions.DELETE_ACCOUNT_FAILURE;
  payload: Error;
}

export interface ClearInfoStatuses {
  type: typeof UserActions.CLEAR_INFO_STATUSES;
  payload: null;
}

export interface ClearPasswordStatuses {
  type: typeof UserActions.CLEAR_PASSWORD_STATUSES;
  payload: null;
}

export interface GetOtherUserStart {
  type: typeof UserActions.GET_OTHER_USER_START;
  payload: string;
}

export interface GetOtherUserSuccess {
  type: typeof UserActions.GET_OTHER_USER_SUCCESS;
  payload: User;
}

export interface GetOtherUserFailure {
  type: typeof UserActions.GET_OTHER_USER_FAILURE;
  payload: Error;
}

export interface ClearOtherUserInfo {
  type: typeof UserActions.CLEAR_OTHER_USER_INFO;
  payload: null;
}

export type UserActionTypes =
  | SignUpStart
  | SignUpSuccess
  | SignUpFailure
  | SignInStart
  | SignInSuccess
  | SignInFailure
  | SignOutStart
  | SignOutSuccess
  | SignOutFailure
  | SignOutFailure
  | CheckUserSession
  | SetCurrentUser
  | ChangeInfoStart
  | ChangeInfoSuccess
  | ChangeInfoFailure
  | ChangePasswordStart
  | ChangePasswordSuccess
  | ChangePasswordFailure
  | ForgotPasswordStart
  | ForgotPasswordSuccess
  | ForgotPasswordFailure
  | ResetPasswordStart
  | ResetPasswordSuccess
  | ResetPasswordFailure
  | DeleteAccountStart
  | DeleteAccountSuccess
  | DeleteAccountFailure
  | ClearInfoStatuses
  | ClearPasswordStatuses
  | GetOtherUserStart
  | GetOtherUserSuccess
  | GetOtherUserFailure
  | ClearOtherUserInfo;
