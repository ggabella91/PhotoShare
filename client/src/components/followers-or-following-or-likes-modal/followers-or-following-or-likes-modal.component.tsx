import React, { useState, useEffect, useRef } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { createStructuredSelector } from 'reselect';
import { AppState } from '../../redux/root-reducer';
import { List } from 'immutable';

import {
  User,
  OtherUserType,
  OtherUserRequest,
} from '../../redux/user/user.types';
import {
  selectFollowersInfo,
  selectFollowingInfo,
} from '../../redux/user/user.selectors';
import { getOtherUserStart } from '../../redux/user/user.actions';

import {
  FileRequestType,
  PostFileReq,
  PostFile,
  UserType,
} from '../../redux/post/post.types';
import {
  selectFollowPhotoFileArray,
  selectUsersProfilePhotoConfirm,
} from '../../redux/post/post.selectors';
import {
  clearFollowPhotoFileArray,
  getPostFileStart,
} from '../../redux/post/post.actions';

import { Follower } from '../../redux/follower/follower.types';

import UserInfo, {
  StyleType,
  UserInfoAndOtherData,
} from '../user-info/user-info.component';

import './followers-or-following-or-likes-modal.styles.scss';
import Modal from 'react-bootstrap/Modal';

interface FollowersOrFollowingOrLikesModalProps {
  users: Follower[] | null;
  show: boolean;
  onHide: () => void;
  isFollowersModal: boolean;
  isPostLikingUsersModal?: boolean;
  postLikingUsersList?: List<UserInfoAndOtherData>;
  followers: User[] | null;
  following: User[] | null;
  followPhotoFileArray: PostFile[] | null;
  usersProfilePhotoConfirm: string | null;
  getOtherUserStart: typeof getOtherUserStart;
  getPostFileStart: typeof getPostFileStart;
  clearFollowPhotoFileArray: typeof clearFollowPhotoFileArray;
}

export interface UserInfoData {
  profilePhotoFileString: string;
  username: string;
  name: string;
  photo: string | null;
  location: string;
  comment: string;
}

export const FollowersOrFollowingOrLikesModal: React.FC<FollowersOrFollowingOrLikesModalProps> =
  ({
    users,
    isFollowersModal,
    isPostLikingUsersModal,
    postLikingUsersList,
    onHide,
    followers,
    following,
    followPhotoFileArray,
    usersProfilePhotoConfirm,
    getOtherUserStart,
    getPostFileStart,
    clearFollowPhotoFileArray,
    ...props
  }) => {
    const [userInfoAndPhotoList, setUserInfoAndPhotoList] = useState<
      List<UserInfoData>
    >(List());

    let usersLoaded = useRef(false);

    let bucket: string;

    if (process.env.NODE_ENV === 'production') {
      bucket = 'photo-share-app-profile-photos';
    } else {
      bucket = 'photo-share-app-profile-photos-dev';
    }

    useEffect(() => {
      if (users && users.length && !usersLoaded.current) {
        usersLoaded.current = true;
        clearFollowPhotoFileArray();

        if (isFollowersModal) {
          users.forEach((user) => {
            getOtherUserStart({
              type: OtherUserType.FOLLOWERS,
              usernameOrId: user.followerId,
            });
          });
        } else {
          users.forEach((user) => {
            getOtherUserStart({
              type: OtherUserType.FOLLOWING,
              usernameOrId: user.userId,
            });
          });
        }
      }
    }, [users]);

    useEffect(() => {
      if (!isPostLikingUsersModal) {
        if (isFollowersModal && followers) {
          handleRenderFollowersOrFollowingInfoArray(followers);
        } else if (!isFollowersModal && following) {
          handleRenderFollowersOrFollowingInfoArray(following);
        }
      }
    }, [followers, following, followPhotoFileArray, usersProfilePhotoConfirm]);

    const handleRenderFollowersOrFollowingInfoArray = (
      followersOrFollowing: User[]
    ) => {
      let followersOrFollowingList: List<User>;

      if (followersOrFollowing.length) {
        followersOrFollowingList = List(followersOrFollowing);
      } else {
        followersOrFollowingList = List();
      }

      if (
        users &&
        followersOrFollowingList.size === users.length &&
        !followPhotoFileArray
      ) {
        followersOrFollowing.forEach((user) => {
          if (user.photo) {
            getPostFileStart({
              user: UserType.followArray,
              s3Key: user.photo,
              bucket,
              fileRequestType: FileRequestType.singlePost,
            });
          }
        });
      } else if (followPhotoFileArray && followPhotoFileArray.length) {
        const followerOrFollowing: List<UserInfoData> =
          followersOrFollowingList.map((el: User) => {
            let photoFileString: string;

            followPhotoFileArray.forEach((file) => {
              if (el.photo === file.s3Key) {
                photoFileString = file.fileString;
              }
            });

            return {
              name: el.name,
              username: el.username,
              profilePhotoFileString: photoFileString!,
              photo: el.photo || '',
              location: '',
              comment: '',
            };
          });

        setUserInfoAndPhotoList(followerOrFollowing);
      } else if (!followPhotoFileArray && usersProfilePhotoConfirm) {
        const followerOrFollowing: List<UserInfoData> =
          followersOrFollowingList.map((el: User) => {
            return {
              name: el.name,
              username: el.username,
              profilePhotoFileString: '',
              photo: el.photo || '',
              location: '',
              comment: '',
            };
          });

        setUserInfoAndPhotoList(followerOrFollowing);
      }
    };

    useEffect(() => {
      if (isPostLikingUsersModal && postLikingUsersList) {
        let postLikerArray: List<UserInfoData>;

        postLikerArray = postLikingUsersList.map((el) => {
          return {
            profilePhotoFileString: el.profilePhotoFileString,
            username: el.username,
            name: el.name,
            photo: '',
            location: '',
            comment: '',
          };
        });

        setUserInfoAndPhotoList(postLikerArray);
      }
    }, [postLikingUsersList]);

    return (
      <Modal
        {...props}
        dialogClassName='followers-following-modal'
        animation={false}
        onHide={onHide}
        centered
        data-testid='followers-following-or-likes-modal'
      >
        <Modal.Header className='followers-following-modal-header' closeButton>
          <span className='header-text'>
            {isPostLikingUsersModal
              ? 'Likes'
              : isFollowersModal
              ? 'Followers'
              : 'Following'}
          </span>
        </Modal.Header>
        <Modal.Body className='followers-following-modal-body'>
          <UserInfo
            userInfoList={userInfoAndPhotoList}
            styleType={StyleType.modal}
          />
        </Modal.Body>
      </Modal>
    );
  };

interface LinkStateProps {
  followers: User[] | null;
  following: User[] | null;
  followPhotoFileArray: PostFile[] | null;
  usersProfilePhotoConfirm: string | null;
}

const mapStateToProps = createStructuredSelector<AppState, LinkStateProps>({
  followers: selectFollowersInfo,
  following: selectFollowingInfo,
  followPhotoFileArray: selectFollowPhotoFileArray,
  usersProfilePhotoConfirm: selectUsersProfilePhotoConfirm,
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  getOtherUserStart: (otherUserReq: OtherUserRequest) =>
    dispatch(getOtherUserStart(otherUserReq)),
  getPostFileStart: (postFileReq: PostFileReq) =>
    dispatch(getPostFileStart(postFileReq)),
  clearFollowPhotoFileArray: () => dispatch(clearFollowPhotoFileArray()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(FollowersOrFollowingOrLikesModal);
