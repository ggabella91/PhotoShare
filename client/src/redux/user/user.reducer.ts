import { Reducer } from 'redux';
import { UserActions, UserActionTypes, UserState } from './user.types';

const INITIAL_STATE: UserState = {
  currentUser: null,
  signUpError: null,
  signInOrOutError: null,
};

const userReducer = (state = INITIAL_STATE, action: UserActionTypes) => {
  switch (action.type) {
    case UserActions.SET_CURRENT_USER:
      return {
        ...state,
        currentUser: action.payload,
      };
    case UserActions.SIGN_UP_SUCCESS:
      return {
        ...state,
        currentUser: action.payload,
        signUpError: null,
      };
    case UserActions.SIGN_IN_SUCCESS:
      return {
        ...state,
        currentUser: action.payload,
        signInOrOutError: null,
      };
    case UserActions.SIGN_OUT_SUCCESS:
      return {
        ...state,
        currentUser: action.payload,
        signInOrOutError: null,
      };
    case UserActions.SIGN_UP_FAILURE:
      return {
        ...state,
        signUpError: action.payload,
      };
    case UserActions.SIGN_IN_FAILURE:
    case UserActions.SIGN_OUT_FAILURE:
      return {
        ...state,
        signInOrOutError: action.payload,
      };
    default:
      return state;
  }
};

export default userReducer;
