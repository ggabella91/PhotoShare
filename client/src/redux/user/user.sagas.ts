import { takeLatest, put, all, call } from 'redux-saga/effects';
import { SagaIterator } from '@redux-saga/core';
import {ActionPattern} from '@redux-saga/types'

import { UserSignUp, User, UserActions, UserActionTypes } from './user.types';

import axios from 'axios';

import {
  setCurrentUser,
  signInSuccess,
  signInFailure,
  signOutSuccess,
  signOutFailure,
  signUpSuccess,
  signUpFailure,
} from './user.actions';

export function* signUp({
  payload: { name, email, password, passwordConfirm },
}: {
  UserActionTypes;
}): SagaIterator {
  try {
    const { data } = yield axios.post('api/v1/users/signup', {
      name,
      email,
      password,
      passwordConfirm,
    });

    yield put(signUpSuccess(data.data.user));
  } catch (err) {
    yield put(signUpFailure(err));
  }
}

export function* onSignUpStart(): SagaIterator {
  yield takeLatest<ActionPattern>(UserActions.SIGN_UP_START, signUp);
}

export function* onSignUpSuccess(): SagaIterator {
  yield takeLatest(UserActions.SIGN_UP_SUCCESS, signInAfterSignUp);
}

export function* signInAfterSignUp({ payload: user }): SagaIterator {
  yield put(setCurrentUser(user));
}

export function* onSignInStart(): SagaIterator {
  yield takeLatest(UserActions.SIGN_IN_START, signIn);
}

export function* onCheckUserSession(): SagaIterator {
  yield takeLatest(UserActions.CHECK_USER_SESSION, isLoggedIn);
}

export function* onSignOutStart(): SagaIterator {
  yield takeLatest(UserActions.SIGN_OUT_START, signOut);
}

export function* userSagas(): SagaIterator {
  yield all([]);
}
