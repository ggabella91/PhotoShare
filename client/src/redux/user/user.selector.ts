import { createSelector } from 'reselect';

import { UserState } from './user.reducer';

const selectUser = (state: any) => state.user;

export const selectCurrentUser = createSelector(
  [selectUser],
  (user: UserState) => user.currentUser
);

export const selectUserSignUpError = createSelector(
  [selectUser],
  (user: UserState) => user.signUpError
);

export const selectUserSignInOrOutError = createSelector(
  [selectUser],
  (user: UserState) => user.signInOrOutError
);
