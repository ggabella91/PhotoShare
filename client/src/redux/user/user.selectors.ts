import { createSelector } from 'reselect';

import { AppState } from '../root-reducer';
import { UserState } from './user.types';

const selectUser = (state: AppState): UserState => state.user;

export const selectCurrentUser = createSelector(
  [selectUser],
  (user: UserState) => user.currentUser
);

export const selectOtherUser = createSelector(
  [selectUser],
  (user: UserState) => user.otherUser
);

export const selectFollowersInfo = createSelector(
  [selectUser],
  (user: UserState) => user.followersInfo
);

export const selectFollowingInfo = createSelector(
  [selectUser],
  (user: UserState) => user.followingInfo
);

export const selectOtherUserError = createSelector(
  [selectUser],
  (user: UserState) => user.otherUserError
);

export const selectPostReactingUsers = createSelector(
  [selectUser],
  (user: UserState) => user.postReactingUsers
);

export const selectUserSuggestions = createSelector(
  [selectUser],
  (user: UserState) => user.userSuggestions
);

export const selectUserSuggestionsConfirm = createSelector(
  [selectUser],
  (user: UserState) => user.userSuggestionsConfirm
);

export const selectUserSuggestionsError = createSelector(
  [selectUser],
  (user: UserState) => user.userSuggestionsError
);

export const selectUserSignUpError = createSelector(
  [selectUser],
  (user: UserState) => user.signUpError
);

export const selectUserSignInOrOutError = createSelector(
  [selectUser],
  (user: UserState) => user.signInOrOutError
);

export const selectChangeInfoConfirm = createSelector(
  [selectUser],
  (user: UserState) => user.changeInfoConfirm
);

export const selectChangeInfoError = createSelector(
  [selectUser],
  (user: UserState) => user.changeInfoError
);

export const selectChangePasswordConfirm = createSelector(
  [selectUser],
  (user: UserState) => user.changePasswordConfirm
);

export const selectChangePasswordError = createSelector(
  [selectUser],
  (user: UserState) => user.changePasswordError
);

export const selectForgotError = createSelector(
  [selectUser],
  (user: UserState) => user.forgotError
);

export const selectForgotConfirm = createSelector(
  [selectUser],
  (user: UserState) => user.forgotConfirm
);

export const selectResetError = createSelector(
  [selectUser],
  (user: UserState) => user.resetError
);

export const selectResetConfirm = createSelector(
  [selectUser],
  (user: UserState) => user.resetConfirm
);

export const selectIsCurrentUserProfilePage = createSelector(
  [selectUser],
  (user: UserState) => user.isCurrentUserProfilePage
);
