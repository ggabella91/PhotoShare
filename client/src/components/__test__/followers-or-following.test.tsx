import { shallow } from 'enzyme';
import React from 'react';

import { getOtherUserStart } from '../../redux/user/user.actions';
import {
  getPostFileStart,
  clearUsersPhotoFileArray,
} from '../../redux/post/post.actions';

import { FollowersOrFollowingModal } from '../followers-or-following-or-likes-modal/followers-or-following-or-likes-modal.component';

it('renders a user-suggestions component', () => {
  const followersOrFollowingModalWrapper = shallow(
    <FollowersOrFollowingModal
      users={[]}
      show={true}
      onHide={() => {}}
      isFollowersModal={true}
      followersOrFollowing={[]}
      usersProfilePhotoArray={[]}
      usersProfilePhotoConfirm=''
      getOtherUserStart={(otherUserReq) => getOtherUserStart(otherUserReq)}
      getPostFileStart={(postFileReq) => getPostFileStart(postFileReq)}
      clearUsersPhotoFileArray={() => clearUsersPhotoFileArray()}
    />
  );

  expect(followersOrFollowingModalWrapper).toMatchSnapshot();
});
