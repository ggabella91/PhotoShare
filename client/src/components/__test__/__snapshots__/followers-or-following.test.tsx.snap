// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a user-suggestions component 1`] = `
<Modal
  animation={false}
  autoFocus={true}
  backdrop={true}
  centered={true}
  clearUsersPhotoFileArray={[Function]}
  dialogAs={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    }
  }
  dialogClassName="followers-following-modal"
  enforceFocus={true}
  followersOrFollowing={Array []}
  keyboard={true}
  onHide={[Function]}
  restoreFocus={true}
  show={true}
  usersProfilePhotoArray={Array []}
>
  <ModalHeader
    className="followers-following-modal-header"
    closeButton={true}
    closeLabel="Close"
  >
    <span
      className="header-text"
    >
      Followers
    </span>
  </ModalHeader>
  <ModalBody
    className="followers-following-modal-body"
  >
    <UserInfo
      styleType="modal"
      userInfoArray={Array []}
    />
  </ModalBody>
</Modal>
`;
